File: sdks/ts/memory/src/connector/notion-properties.ts
// SPDX-License-Identifier: Apache-2.0

/**
 * Property extraction helpers for the Notion connector. Converts
 * Notion property objects to their string representations and
 * parses database entries into structured data.
 */

import { renderPlainRichText, type NotionRichText } from './notion-blocks.js'

// ---------- Types ----------

/**
 * Parsed database entry.
 */
export type ParsedDatabaseEntry = {
  readonly pageId: string
  readonly title: string
  readonly url: string
  readonly lastEditedTime: Date
  readonly propertiesMarkdown: string
}

/**
 * Parsed page metadata.
 */
export type ParsedPage = {
  readonly id: string
  readonly title: string
  readonly url: string
  readonly lastEditedTime: Date
}

// ---------- Title extraction ----------

/**
 * Extracts the title from a page's properties map.
 */
export const extractTitleFromProperties = (
  properties: Record<string, Record<string, unknown>>,
): string => {
  for (const prop of Object.values(properties)) {
    if (prop.type === 'title') {
      const titleArr = (prop.title ?? []) as readonly NotionRichText[]
      return renderPlainRichText(titleArr)
    }
  }
  return ''
}

// ---------- Property value extraction ----------

/**
 * Converts a Notion property to its string representation.
 */
export const extractPropertyValue = (
  prop: Record<string, unknown>,
): string => {
  const propType = prop.type as string

  const renderers: Readonly<Record<string, () => string>> = {
    title: () =>
      renderPlainRichText((prop.title ?? []) as readonly NotionRichText[]),
    rich_text: () =>
      renderPlainRichText(
        (prop.rich_text ?? []) as readonly NotionRichText[],
      ),
    number: () => {
      const num = prop.number as number | null
      return num !== null && num !== undefined ? String(num) : ''
    },
    select: () => {
      const sel = prop.select as { name: string } | null
      return sel?.name ?? ''
    },
    multi_select: () => {
      const items = (prop.multi_select ?? []) as readonly { name: string }[]
      return items.map((i) => i.name).join(', ')
    },
    date: () => {
      const d = prop.date as { start: string; end?: string } | null
      if (d === null || d === undefined) return ''
      return d.end !== undefined && d.end !== ''
        ? `${d.start} to ${d.end}`
        : d.start
    },
    checkbox: () => (prop.checkbox === true ? 'true' : 'false'),
    url: () => (prop.url as string) ?? '',
    email: () => (prop.email as string) ?? '',
    phone_number: () => (prop.phone_number as string) ?? '',
    status: () => {
      const st = prop.status as { name: string } | null
      return st?.name ?? ''
    },
    people: () => {
      const people = (prop.people ?? []) as readonly { name: string }[]
      return people.map((p) => p.name).join(', ')
    },
    relation: () => {
      const rels = (prop.relation ?? []) as readonly { id: string }[]
      return rels.map((r) => r.id).join(', ')
    },
  }

  const renderer = renderers[propType]
  return renderer !== undefined ? renderer() : ''
}

// ---------- Database entry parsing ----------

/**
 * Parses a database entry from the query response.
 */
export const parseDatabaseEntry = (
  raw: Record<string, unknown>,
): ParsedDatabaseEntry => {
  const properties = (raw.properties ?? {}) as Record<
    string,
    Record<string, unknown>
  >
  const lastEditedStr = raw.last_edited_time as string
  const lastEditedTime = new Date(lastEditedStr)

  let title = ''
  const propLines: string[] = []

  // Sort keys for deterministic output.
  const sortedKeys = Object.keys(properties).sort()

  for (const key of sortedKeys) {
    const prop = properties[key]
    if (prop === undefined) continue
    const val = extractPropertyValue(prop)

    if (prop.type === 'title') {
      title = val
      continue
    }

    if (val !== '') {
      propLines.push(`- **${key}**: ${val}`)
    }
  }

  let markdown = ''
  if (title !== '') {
    markdown += `## ${title}\n\n`
  }
  if (propLines.length > 0) {
    markdown += propLines.join('\n') + '\n'
  }

  return {
    pageId: raw.id as string,
    title,
    url: raw.url as string,
    lastEditedTime,
    propertiesMarkdown: markdown,
  }
}

// ---------- Page response parsing ----------

/**
 * Parses a Notion API page response into structured metadata.
 */
export const parsePageResponse = (
  data: Record<string, unknown>,
): ParsedPage => {
  const properties = (data.properties ?? {}) as Record<
    string,
    Record<string, unknown>
  >
  const title = extractTitleFromProperties(properties)
  const lastEditedStr = data.last_edited_time as string
  const lastEditedTime = new Date(lastEditedStr)

  return {
    id: data.id as string,
    title,
    url: data.url as string,
    lastEditedTime,
  }
}

File: go/connector/notion_properties.go
// SPDX-License-Identifier: Apache-2.0

package connector

import (
	"encoding/json"
	"fmt"
	"sort"
	"strings"
	"time"
)

// extractTitleFromProperties finds the title property in a page's
// properties map and returns its plain text value.
func extractTitleFromProperties(properties map[string]json.RawMessage) string {
	for _, propRaw := range properties {
		var prop struct {
			Type  string `json:"type"`
			Title []struct {
				PlainText string `json:"plain_text"`
			} `json:"title"`
		}
		if err := json.Unmarshal(propRaw, &prop); err != nil {
			continue
		}
		if prop.Type == "title" && len(prop.Title) > 0 {
			var parts []string
			for _, t := range prop.Title {
				parts = append(parts, t.PlainText)
			}
			return strings.Join(parts, "")
		}
	}
	return ""
}

// extractPropertyValue converts a Notion property value to its
// string representation.
func extractPropertyValue(raw json.RawMessage) string {
	var prop struct {
		Type           string           `json:"type"`
		Title          []notionRichText `json:"title"`
		RichText       []notionRichText `json:"rich_text"`
		Number         *float64         `json:"number"`
		Select         *struct {
			Name string `json:"name"`
		} `json:"select"`
		MultiSelect []struct {
			Name string `json:"name"`
		} `json:"multi_select"`
		Date *struct {
			Start string `json:"start"`
			End   string `json:"end"`
		} `json:"date"`
		Checkbox bool   `json:"checkbox"`
		URL      string `json:"url"`
		Email    string `json:"email"`
		Phone    string `json:"phone_number"`
		Status   *struct {
			Name string `json:"name"`
		} `json:"status"`
		People []struct {
			Name string `json:"name"`
		} `json:"people"`
		Relation []struct {
			ID string `json:"id"`
		} `json:"relation"`
	}
	if err := json.Unmarshal(raw, &prop); err != nil {
		return ""
	}

	renderers := map[string]func() string{
		"title":    func() string { return renderPlainRichText(prop.Title) },
		"rich_text": func() string { return renderPlainRichText(prop.RichText) },
		"number": func() string {
			if prop.Number == nil {
				return ""
			}
			return fmt.Sprintf("%g", *prop.Number)
		},
		"select": func() string {
			if prop.Select == nil {
				return ""
			}
			return prop.Select.Name
		},
		"multi_select": func() string {
			names := make([]string, 0, len(prop.MultiSelect))
			for _, ms := range prop.MultiSelect {
				names = append(names, ms.Name)
			}
			return strings.Join(names, ", ")
		},
		"date": func() string {
			if prop.Date == nil {
				return ""
			}
			if prop.Date.End != "" {
				return prop.Date.Start + " to " + prop.Date.End
			}
			return prop.Date.Start
		},
		"checkbox": func() string {
			if prop.Checkbox {
				return "true"
			}
			return "false"
		},
		"url":          func() string { return prop.URL },
		"email":        func() string { return prop.Email },
		"phone_number": func() string { return prop.Phone },
		"status": func() string {
			if prop.Status == nil {
				return ""
			}
			return prop.Status.Name
		},
		"people": func() string {
			names := make([]string, 0, len(prop.People))
			for _, p := range prop.People {
				names = append(names, p.Name)
			}
			return strings.Join(names, ", ")
		},
		"relation": func() string {
			ids := make([]string, 0, len(prop.Relation))
			for _, r := range prop.Relation {
				ids = append(ids, r.ID)
			}
			return strings.Join(ids, ", ")
		},
	}

	renderer, found := renderers[prop.Type]
	if !found {
		return ""
	}
	return renderer()
}

// databaseEntryParsed holds parsed fields from a database entry.
type databaseEntryParsed struct {
	pageID             string
	title              string
	url                string
	lastEditedTime     time.Time
	propertiesMarkdown string
}

// parseDatabaseEntry extracts structured data from a raw database
// entry JSON.
func parseDatabaseEntry(raw json.RawMessage) (databaseEntryParsed, error) {
	var entry struct {
		ID             string                     `json:"id"`
		URL            string                     `json:"url"`
		LastEditedTime string                     `json:"last_edited_time"`
		Properties     map[string]json.RawMessage `json:"properties"`
	}
	if err := json.Unmarshal(raw, &entry); err != nil {
		return databaseEntryParsed{}, err
	}

	editedTime, _ := time.Parse(time.RFC3339, entry.LastEditedTime)

	title := ""
	var propLines []string

	// Sort property keys for deterministic output.
	keys := make([]string, 0, len(entry.Properties))
	for k := range entry.Properties {
		keys = append(keys, k)
	}
	sort.Strings(keys)

	for _, key := range keys {
		propRaw := entry.Properties[key]
		val := extractPropertyValue(propRaw)

		// Detect title property.
		var propType struct {
			Type string `json:"type"`
		}
		if err := json.Unmarshal(propRaw, &propType); err == nil && propType.Type == "title" {
			title = val
			continue
		}

		if val != "" {
			propLines = append(propLines, fmt.Sprintf("- **%s**: %s", key, val))
		}
	}

	var builder strings.Builder
	if title != "" {
		builder.WriteString("## " + title + "\n\n")
	}
	if len(propLines) > 0 {
		builder.WriteString(strings.Join(propLines, "\n"))
		builder.WriteString("\n")
	}

	return databaseEntryParsed{
		pageID:             entry.ID,
		title:              title,
		url:                entry.URL,
		lastEditedTime:     editedTime,
		propertiesMarkdown: builder.String(),
	}, nil
}

// parsePageResponse extracts page metadata from a Notion page API
// response.
func parsePageResponse(data []byte) (NotionPage, error) {
	var resp struct {
		ID             string                     `json:"id"`
		URL            string                     `json:"url"`
		LastEditedTime string                     `json:"last_edited_time"`
		Properties     map[string]json.RawMessage `json:"properties"`
		Parent         struct {
			Type   string `json:"type"`
			PageID string `json:"page_id"`
		} `json:"parent"`
	}
	if err := json.Unmarshal(data, &resp); err != nil {
		return NotionPage{}, fmt.Errorf("parsing page: %w", err)
	}

	editedTime, _ := time.Parse(time.RFC3339, resp.LastEditedTime)

	// Extract title from properties.
	title := extractTitleFromProperties(resp.Properties)

	return NotionPage{
		ID:             resp.ID,
		Title:          title,
		URL:            resp.URL,
		LastEditedTime: editedTime,
		ParentID:       resp.Parent.PageID,
	}, nil
}
